Several editorial changes and rewrites (#18)
cfrg/draft-irtf-cfrg-hybrid-kems: draft-irtf-cfrg-hybrid-kems.md

  FIPS203: DOI.10.6028/NIST.FIPS.203

informative:
  ABR01:
    title: "The Oracle Diffie-Hellman Assumptions and an Analysis of DHIES"
    date: Jan, 2001
    author:
      -
        ins: Michel Abdalla
      -

>         ins: Mihir Bellare1

Reviewer: Typo

--- abstract

> "Post-quantum" (PQ) algorithms promise to resist attack by a quantum computer,

Reviewer: Feel free to ignore this stylistic blabbering: personally I like to write "post-quantum algorithms are designed to resist attack of quantum computers". Then we don't need any qualifications like "promise" or "believe to" etc. Of course it's not completely true: several schemes were first designed before quantum computers were known about.

in contrast to "traditional" algorithms.  However, given the novelty of PQ
algorithms, there is some concern that PQ algorithms currently believed to be

> secure will be broken.  Hybrid constructions that combine both PQ and

Author: Added a couple of sentences to this effect in the Introduction.

properties as long as the undelying algorithms are secure.

--- middle

# Introduction {#intro}

Post-quantum (PQ) algorithms offer a redesign of traditional algorithms tailored
towards resisting attack from a quantum computer. Key Encapsulation Mechanisms
(KEMs), are a standardized algorithm type that can be used to build protocols in
lieu of traditional, quantum-vulnerable variants such as Diffie-Hellman (DH)
based protocols.  Upgrading protocols to use PQ KEMs is a priority for the
protocol design community, due to the possibility of "harvest now, decrypt
later" attacks.

Given the novelty of PQ algorithms, however, there is some concern that PQ
algorithms currently believed to be secure will be broken.  Hybrid
constructions that combine both PQ and traditional algorithms can help moderate
this risk while still providing security against quantum attack.  If construted
properly, a hybrid KEM will retain the properties of either constituent KEM
even if the other KEM is compromised.  If the PQ KEM is broken by new
cryptanalysis, then the hybrid KEM should continue to provide security against
non-quantum attackers by virtue of its traditional KEM component.  If the

> traditional KEM is brken by a quantum computer, then the hybrid KEM should

Reviewer: `brken` typo

--- abstract

"Post-quantum" (PQ) algorithms promise to resist attack by a quantum computer,
in contrast to "traditional" algorithms.  However, given the novelty of PQ
algorithms, there is some concern that PQ algorithms currently believed to be
secure will be broken.  Hybrid constructions that combine both PQ and
traditional algorithms can help moderate this risk while still providing
security against quantum attack.  In this document, we define constructions for
hybrid Key Encapsulation Mechanisms (KEMs) based on combining a traditional KEM
and a PQ KEM.  Hybrid KEMs using these constructions provide strong security
properties as long as the undelying algorithms are secure.

--- middle

# Introduction {#intro}

"Post-quantum" (PQ) algorithms promise to resist attack by a quantum computer,
in contrast to "traditional" algorithms.  Key Encapsulation Mechanisms (KEMs)
are an area of particular concern.  As of this writing, it is unlikely that a

> quantum computer already exists, but it is possible that one might be created in

Reviewer 2: I recommend that we avoid "timed" statements in the draft that might not age well. This can be changed to:

Post-quantum (PQ) algorithms offer a redesign of traditional algorithms tailored towards resisting attack from a quantum computer. Key Encapsulation Mechanisms (KEMs), are a standardized algorithm type that can be used to build protocols in lieu of traditional, quantum-vulnerable variants such as Diffie-Hellman (DH) based protocols. Given the novelty of PQ algorithms, however....

--- abstract

"Post-quantum" (PQ) algorithms are designed to resist attack by a quantum computer,
in contrast to "traditional" algorithms.  However, given the novelty of PQ
algorithms, there is some concern that PQ algorithms currently believed to be
secure will be broken.  Hybrid constructions that combine both PQ and
traditional algorithms can help moderate this risk while still providing
security against quantum attack. In this document, we define constructions for
hybrid Key Encapsulation Mechanisms (KEMs) based on combining a traditional KEM
and a PQ KEM. Hybrid KEMs using these constructions provide strong security
properties as long as the undelying algorithms are secure.

--- middle

# Introduction {#intro}

> Post-quantum (PQ) algorithms offer a redesign of traditional algorithms tailored
> towards resisting attack from a quantum computer. Key Encapsulation Mechanisms
> (KEMs), are a standardized algorithm type that can be used to build protocols in

Reviewer 3: ```suggestion
Post-quantum (PQ) algorithms offer new constructions based on problems tailored
towards resisting attack from a quantum computer. Key Encapsulation Mechanisms
(KEMs), are a standardized algorithm type that can be used to build protocols in
```

> lieu of traditional, quantum-vulnerable variants such as Diffie-Hellman (DH)
> based protocols.  Upgrading protocols to use PQ KEMs is a priority for the

Reviewer 3: ```suggestion
lieu of traditional, quantum-vulnerable variants such as finite field
or elliptic curve Diffie-Hellman (DH) based protocols. Upgrading protocols
to use PQ KEMs is a priority for the
```

> based protocols.  Upgrading protocols to use PQ KEMs is a priority for the
> protocol design community, due to the possibility of "harvest now, decrypt
> later" attacks.

Reviewer 3: ```suggestion
based protocols.  Upgrading key establishment protocols to use PQ KEMs is a
priority for the protocol design community, due to the possibility of
"harvest now, decrypt later" attacks.
```

Given the novelty of PQ algorithms, however, there is some concern that PQ
algorithms currently believed to be secure will be broken.  Hybrid
constructions that combine both PQ and traditional algorithms can help moderate
this risk while still providing security against quantum attack.  If construted
properly, a hybrid KEM will retain the properties of either constituent KEM

> even if the other KEM is compromised.  If the PQ KEM is broken by new

Reviewer 3: 'retains the properties' needs clarity

cryptanalysis, then the hybrid KEM should continue to provide security against
non-quantum attackers by virtue of its traditional KEM component.  If the
traditional KEM is broken by a quantum computer, then the hybrid KEM should
continue to resist quantum attack by virtue of its PQ KEM component.

> In addition to guarding against algorithm flaws, this property also guards
> against flaws in implementations, such as timing attacks.  Hybrid KEMs can also
> facilitate faster deployment of PQ security by allowing applications to

Reviewer 3: ```suggestion
In addition to guarding against algorithm weakness, this property also guards
against flaws in implementations, such as timing attacks.  Hybrid KEMs can also
facilitate faster deployment of PQ security by allowing applications to
```

incorporate PQ algorithms while still meeting compliance requirements based on
traditional algorithms.

In this document, we define constructions for hybrid KEMs based on combining a

> traditional KEM and a PQ KEM.  The aim of this document is provide a small set

Reviewer 3: But not only KEMs as components, as in QSF - we combine traditional components and PQ KEMs

if `x = [0, 1, 2, 3, 4]`, then `x[..2] = [0, 1]` and `x[2..] = [2, 3, 4]`.

# Cryptographic Dependencies {#cryptographic-deps}

The generic hybrid PQ/T KEM constructions we define depend on the the
following cryptographic primitives:

- Key Encapsulation Mechanism {{kems}}
- Extendable Output Function (XOF) {{xof}}
- Key Derivation Function (KDF) {{kdf}}

In the remainder of this section, we describe functional aspects of these
mechanisms.  The security properties we require in order for the resulting
hybrid KEM to be secure are discussed in {{security-properties}}.

## Key encapsulation mechanisms {#kems}

~~~ aasvg
     +-----------------+
     | GenerateKeyPair |
     |        or       |
     |  DeriveKeyPair  |
     +--------+--------+
              |
    +---------+----------+
    |                    |
    V                    V

    ek                  dk

    |                    |
    |                    |
    V                    V
+--------+    ct    +--------+
| Encaps |--------->| Decaps |
+--------+          +--------+
    |                    |
    |                    |
    V                    V

    ss        ==        ss

> ~~~

Reviewer 3: nice

A Key Encapsulation Mechanism (KEMs) comprises the following algorithms:

- `GenerateKeyPair() -> (ek, dk)`: A randomized algorithm that generates a
  public encapsulation key `ek` and a secret decapsulation key `dk`, each of
  which are byte strings.
- `DeriveKeyPair(seed) -> (ek, dk)`: A deterministic algorithm that takes as
  input a seed `seed` and generates a public encapsulation key `ek` and a secret
  decapsulation key `dk`, each of which are byte strings.
- `Encaps(ek) -> (ct, ss)`: A probabilistic encapsulation
  algorithm, which takes as input a public encapsulation key `ek` and outputs
  a ciphertext `ct` and shared secret `ss`.
- `Decaps(dk, ct) -> ss`: A decapsulation algorithm, which takes
  as input a secret decapsulation key `dk` and ciphertext `ct` and outputs a

>   shared secret `ss`.

Reviewer 3: 'Yes but', in the literature there is just keygen, so idk if we want to add a word or two to like, make this more 'applied', or applicable just in this document, or something

A KEM may also provide a deterministic version of `Encaps` (e.g., for purposes
of testing):

- `EncapsDerand(ek, randomness) -> (ct, shared_secret)`: A deterministic
   encapsulation algorithm, which takes as input a public encapsulation key
   `ek` and randomness `randomness`, and outputs a ciphertext `ct` and shared
   secret `shared_secret`.

> We assume that the values produced and consumed by the above functions are all
> byte strings, with fixed lengths:

Reviewer 3: ```suggestion
We assume that the values produced and consumed by the above functions are all
byte strings, with fixed lengths per parameter set:
```

- `Nseed`: The length of a key seed (input to DeriveKeyPair)
- `Nek`: The length of a public encapsulation key
- `Ndk`: The length of a secret decapsulation key
- `Nct`: The length of a ciphertext produced by Encaps
- `Nss`: The length of a shared secret produced by Encaps or Decaps

This interface is effectively the same as the one defined in the Hybrid Public
Key Encryption (HPKE) specification {{?RFC9180}}.  The only difference is that
here we assume that all values are byte strings, whereas in HPKE keys are opaque
by default and serialized or deserialized as needed.  We also use slightly
different terminology for keys, emphasizing "encapsulation" and "decapsulation"

> as opposed to "public" and "secret".

Reviewer 3: True, but why mention it?

Author: In case people have already read HPKE and are looking at this.  I have added markup to make it an inset note / aside.  I'm not that committed to it; happy to delete it if people feel strongly.

pseudorandom function (PRF) in the standard model {{GHP2018}} and independent
random oracle in the random oracle model (ROM).

## KEM from Diffie-Hellman {#group}

This section describes a simple KEM built from a Diffie-Hellman group.  **This

> KEM is not IND-CCA secure**. It is, however IND-CPA secure under either the

Reviewer: IND-CCA and IND-CPA might be meaningless to the reader at this point. What about the following

> This KEM is not secure against active attack (not IND-CCA).

Author: Reworded this to be more reader-friendly, and provide a forward pointer to the security considerations.

- `RandomScalar(seed) -> k`: Produce a uniform pseudo-random scalar from the
  byte string `seed`.
- `ScalarToBytes(k) -> dk`: Encode a scalar into a fixed-length byte string.
- `BytesToScalar(dk) -> k`: Decode a scalar from a fixed-length byte string.
- `ElementToBytes(P) -> dk`: Encode an element of the group into a fixed-length
  byte string.
- `BytesToElement(dk) -> P`: Decode an element of the group from a fixed-length
  byte string.
- `ElementToSharedSecret(P) -> ss`: Extract a shared secret from an element of
  the group (e.g., by taking the X coordinate of an ellpitic curve point).

Based on this notion of a group, we can define a DH-based KEM as follows:

~~~
def GenerateKeyPair():
    seed = random(Nseed)
    return DeriveKeyPair(seed)

def DeriveKeyPair(seed):
    p = RandomScalar(seed)
    P = p * B
    dk = ScalarToBytes(p)
    ek = ElementToBytes(P)
    return (ek, dk)

def Encaps(ek):
    P = BytesToElement(ek)
    (Q, q) = GenerateKeyPair()
    ct = ElementToBytes(Q)
    ss = ElementToSharedSecret(q * P)
    return (ct, ss)

def Decaps(dk, ct)
    p = BytesToScalar(dk)
    Q = BytesToElement(ct)
    ss = ElementToSharedSecret(p * Q)
    return ss

> ~~~

Reviewer: I know it makes sense to define the IND-CPA KEM like this, but I'm worried that once it's defined, it's very tempting for implementors to put it in the KEM slot and have it available to users, which do not understand warnings such as "this is only IND-CPA".

Not assigning a code points helps, but maybe we can do more.

Ideally we'd make the interface of a component incompatible with a proper KEM, but I don't see any non-artificial way to do that.

Different naming only goes so far: I think *Component* or say *PreKEM* would be better.

Author: I added some text that explicitly says "This KEM should not be used on its own."  Unfortunately the 🚨 emoji isn't allowed in RFCs.

* `XOF`: the eXtended Output Function
* `PQKEM`: the PQ KEM component scheme
* `G`: the nomimal group used to construct the traditional KEM component
  scheme as described in {{group}}
* Nseed: length in bytes of the seed randomness sourced from the RNG
* Npqseed: length in bytes of the input to PQ.DeriveKeyPair()
* Ntradseed: length in bytes of the input to NominalGroup.ScalarFromBytes()

~~~
def expandDecapsulationKey(dk):
  expanded = XOF(dk, Npqseed + Ntradseed)
  (pq_EK, pq_DK) = PQKEM.DeriveKeyPair(expanded[..Npqseed])
  trad_DK = G.ScalarFromBytes(expanded[Npqseed..])

>   trad_EK = G.SerializeElement(NominalGroup.ScalarMultBase(trad_DK))

Reviewer: This is breaking abstraction.

Author: Yep, this will get fixed in the next PR.  The only reason this part is getting touched is the global `(sk, pk) -> (dk, ek)` terminology change.
